### ISSUE TITLE
Issues using EMR4.3.0

### ISSUE THREAD

**Issue by asicoe** [krux/hyperion#369]
I have a bunch of AWS DataPipelines using EMR4.3.0 with Spark 1.6.0 and naturally I want to programatically define them rather than manually writing up JSON definition files.

I just recently came across the Hyperion project which seems like something I could use. Just by looking at the Spark example and a bit of digging around though, it seems that the library is just supporting older style of data pipelines definitions. 

For example the apply methods of both SparkTaskActivity and SparkStep are instantiating their private companion classes with hard coded values for:
scriptRunner = "s3://elasticmapreduce/libs/script-runner/script-runner.jar",
jobRunner = s"${hc.scriptUri}run-spark-step.sh",

These are pre EMR 4.x.x ways of running spark. In EMR4.x.x it's recommended to use command-runner.jar and spark-submit e.g.:
```
{
 "name":"MyActivity",
 "step":[
  "command-runner.jar,",
  "spark-submit,",
  "--deploy-mode,cluster,",
  ...
 ]
 ...
}
```

I also noticed that in SparkCluster the way to specify Spark is via a standard bootstrap action s3://support.elasticmapreduce/spark/install-spark. The EMR 4.x.x way is to specify "applications":["spark"] in the EMRCluster object. 

There are other differences as well.

Some questions:
- Do you guys have plans to upgrade the DSL to support new EMR versions? When?
- Currently the configs I mentioned above are hard coded behind private case classes which cannot be directly instantiated by code using the library. Am I missing something. Is there another way to extend the functionality to allow generation of definitions for EMR4.x.x? Or do I need to modify the Hyperion code to do that?
- If modifying the Hyperion code is the only way forward, would you guys be willing to accept contributions in this direction?

**Comment by realstraw** [comment]
Hi Alex, 

I'm copying my email reply here for more exposure and completeness.

> - Do you guys have plans to upgrade the DSL to support new EMR versions? When?

Yes, we do plan to have better support for EMR release labels, and I do realize that it has some problem with the current version. We started to address some of them, and I’m planning to add and address Spark issues on EMR release label release. Currently the plan is I will fix it within the next 1 to 2 months if no others address it before me.

> Currently the configs I mentioned above are hard coded behind private case classes which cannot be directly instantiated by code using the library. Am I missing something. Is there another way to extend the functionality to allow generation of definitions for EMR4.x.x? Or do I need to modify the Hyperion code to do that?

As you mentioned the current spark cluster is basically a MapReduceCluster with some additional settings, you can extend it easily by using your own MapReduceStep, with your own MapReduceCluster with spark configured.

> If modifying the Hyperion code is the only way forward, would you guys be willing to accept contributions in this direction?

We welcome any PRs, please try to follow this guideline https://github.com/krux/hyperion/blob/master/CONTRIBUTING.md if you decide to contribute.

**Comment by asicoe** [comment]
Hi and thanks for the reply!

I did manage to generate a definition.json fit for running Spark on EMR 4.3.0 with not much hassle, just had to create my own SparkActivity and SparkCluster largely based on the existing implementations with some tweaks. This is not ideal but works for now.

The only feature that I'm stuck on now is adding custom properties to the default object. This is useful to overcome the max 50 limitation on the number of properties a data pipelines has. I tried overriding the defaultObject method in DataPipelineDef by returning my own class that extends DefaultObject but am running into trouble with case class extension in Scala and it still picks up the superclass implementation. The DefaultObject has a hard coded map.

Is there any way for adding extra properties to the default object without having to change library code?

Thanks
